openedx/openedx-platform
# feat: only log requests that fail user verification in SafeSessionMiddleware

## Description
Update the LOG_REQUEST_USER_CHANGES toggle to have it determine whether or not we will save information about all changes to the user in a request rather than logging it immediately. We will only log this information if the request fails the subsequent verify_user check in the response.
This will only affect Developers.

## Supporting information
https://openedx.atlassian.net/browse/ARCHBOM-1923

## Testing instructions

This guy is hard to test outside a development environment. Inside a development environment, you can test by enabling the LOG_REQUEST_USER_CHANGES toggle and disabling the ```        if getattr(response, 'safe_sessions_expected_user_change', None):
            return``` check inside SafeSessionMiddleware.process_response. 
Then:
1. while logged out, open two login windows
2. log in to one of them (keeping the other window open)
3. in the other window, log in using a different user
In the logs, you should see a message `SafeCookieData user at initial request does not match user at response time` and information about how the SafeCookieData user has changed over the life of the request.

## Deadline

None
